File: src/test/java/no/nav/pensjondokdist/distribuerjournalpost/DistribuerJournalpostClientTest.java
package no.nav.pensjondokdist.distribuerjournalpost;

import static org.springframework.test.web.client.ExpectedCount.manyTimes;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withBadRequest;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.client.MockRestServiceServer;
import org.springframework.web.client.RestTemplate;

import no.nav.pensjondokdist.PensjonDokdistException;
import no.nav.pensjondokdist.distribuerjournalpost.dto.DistribuerJournalpostRequest;
import no.nav.pensjondokdist.oidc.OidcProvider;

@SpringBootTest
@RunWith(MockitoJUnitRunner.class)
public class DistribuerJournalpostClientTest {
    private static final String URL = "http://SafUrl.com";
    private DistribuerJournalpostClient client;
    @Mock
    private OidcProvider provider;
    private RestTemplate restTemplate = new RestTemplate();

    @Before
    public void init() {
        client = new DistribuerJournalpostClient(URL, restTemplate, provider);
    }

    @Test
    public void testGetWithSuccessResponse() {
        MockRestServiceServer server = MockRestServiceServer.bindTo(restTemplate).build();

        server.expect(manyTimes(), requestTo(URL))
                .andRespond(withSuccess("", MediaType.APPLICATION_JSON));
        client.post(new DistribuerJournalpostRequest("12345", null, null, null, null, null, null));
        server.verify();
    }

    @Test(expected = PensjonDokdistException.class)
    public void testGetWithBadRequestResponseThrowException() {
        MockRestServiceServer server = MockRestServiceServer.bindTo(restTemplate).build();
        server.expect(manyTimes(), requestTo(URL))
                .andRespond(withBadRequest());
        client.post(new DistribuerJournalpostRequest("12345", null, null, null, null, null, null));
        server.verify();
    }
}

File: src/test/kotlin/no/nav/pensjon/dokdist/dokdistfordeling/DistribuerJournalpostClientTest.kt
package no.nav.pensjon.dokdist.dokdistfordeling

import com.fasterxml.jackson.databind.ObjectMapper
import no.nav.pensjon.dokdist.DokDistException
import no.nav.pensjon.dokdist.auth.OnBehalfOfTokenResponse
import no.nav.pensjon.dokdist.dokdistfordeling.DistribuerJournalpostRequest.Distribusjonstype
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.*
import org.springframework.boot.test.web.client.MockServerRestTemplateCustomizer
import org.springframework.boot.web.client.RestTemplateBuilder
import org.springframework.http.*
import org.springframework.test.web.client.match.MockRestRequestMatchers.*
import org.springframework.test.web.client.response.MockRestResponseCreators.*

private const val endpoint = "http://dokdistfordeling.local"

class DistribuerJournalpostClientTest {
    private val objectMapper = ObjectMapper()

    private val token = OnBehalfOfTokenResponse("Bearer", "their-scope", "1000", "the-token")
    private val journalpostId = "12345"
    private val distribuerUrl = "$endpoint/rest/v1/distribuerjournalpost"
    private val response = DistribuerJournalpostResponse("991122")
    private val responseJson = objectMapper.writeValueAsString(response)

    private val mockServerCustomizer: MockServerRestTemplateCustomizer = MockServerRestTemplateCustomizer()
    private val distribuer = DistribuerJournalpostClient(endpoint, token, RestTemplateBuilder(mockServerCustomizer))
    private val mockServer = mockServerCustomizer.server

    @Test
    fun `can distribuere journalpost`() {
        val expectedRequest = DistribuerJournalpostRequest(
            journalpostId,
            BESTILLENDE_FAGSYSTEM,
            BESTILLENDE_FAGSYSTEM,
            Distribusjonstype.VEDTAK,
            DistribuerJournalpostRequest.Distribusjonstidspunkt.KJERNETID
        )
        mockServer.expect(requestTo(distribuerUrl))
            .andExpect(method(HttpMethod.POST))
            .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
            .andExpect(content().json(objectMapper.writeValueAsString(expectedRequest)))
            .andRespond(withSuccess(responseJson, MediaType.APPLICATION_JSON))

        assertThat(distribuer.distribuer(journalpostId, expectedRequest.distribusjonstype).originalResponse).isEqualTo(response)
        mockServer.verify()
    }

    @Test
    fun `fails if response is empty`() {
        mockServer.expect(requestTo(distribuerUrl))
            .andRespond(withSuccess())

        assertThrows<DokDistException> { distribuer.distribuer(journalpostId, Distribusjonstype.VEDTAK) }
        mockServer.verify()
    }

    @Test
    fun `handles badrequest`() {
        mockServer.expect(requestTo(distribuerUrl))
            .andRespond(withBadRequest())

        assertThrows<DokDistException> { distribuer.distribuer(journalpostId, Distribusjonstype.VEDTAK) }
        mockServer.verify()
    }

    @Test
    fun `handles server error`() {
        mockServer.expect(requestTo(distribuerUrl))
            .andRespond(withServerError())

        assertThrows<DokDistException> { distribuer.distribuer(journalpostId, Distribusjonstype.VEDTAK) }
        mockServer.verify()
    }

}
